lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,461 @@
+    """
+}
+
+extension DefaultStringInterpolation {
+    /// Generate a _partial_ Swift literal from the given bytes. It is partial in that must be embedded
+    /// into some other syntax, specifically as a tuple.
+    fileprivate mutating func appendInterpolation(swiftLiteral bytes: [UInt8]) {
+        let literalBytes = bytes.map({ "\($0) as UInt8" }).joined(separator: ", ")
+        appendInterpolation(literalBytes)
+    }
+}
+
+extension String {
+    /// Encode a string into UTF8 bytes, prefixed by a ULEB length, and suffixed by the null terminator.
+    fileprivate var byteEncoded: [UInt8] {
+        let size = UInt(self.utf8.count) + 1 // including null terminator
+        var bytes: [UInt8] = []
+        bytes.append(contentsOf: size.ULEBEncoded)
+        bytes.append(contentsOf: self.utf8)
+        bytes.append(0) // null terminator
+        return bytes
+    }
+}
+
+extension UInt {
+    /// Encode an unsigned integer into ULEB format. See https://en.wikipedia.org/wiki/LEB128
+    fileprivate var ULEBEncoded: [UInt8] {
+        guard self > 0 else {
+            return [0]
+        }
+
+        var bytes: [UInt8] = []
+        var buffer = self
+        while buffer > 0 {
+            var byte = UInt8(buffer & 0b0111_1111)
+            buffer >>= 7
+            if buffer > 0 {
+                byte |= 0b1000_0000
+            }
+            bytes.append(byte)
+        }
+        return bytes
+    }
+}
+
+// MARK: - Diagnostics
+
+fileprivate struct ErrorMessage: DiagnosticMessage, ExpressibleByStringInterpolation {
+    init(stringLiteral value: String) {
+        self.message = value
+    }
+    var message: String
+    var diagnosticID: MessageID { .init(domain: "DebugDescription", id: "DebugDescription")}
+    var severity: DiagnosticSeverity { .error }
+}
+
+extension MacroExpansionContext {
+    fileprivate func diagnose(node: some SyntaxProtocol, error message: ErrorMessage) {
+        diagnose(Diagnostic(node: node, message: message))
+    }
+}
+
+// MARK: - Syntax Tree Helpers
+
+extension MacroExpansionContext {
+    /// Determine the module name of the Syntax node, via its fileID.
+    /// See https://developer.apple.com/documentation/swift/fileid()
+    fileprivate func moduleName<T: SyntaxProtocol>(of node: T) -> String? {
+        if let fileID = self.location(of: node)?.file.as(StringLiteralExprSyntax.self)?.representedLiteralValue,
+           let firstSlash = fileID.firstIndex(of: "/") {
+            return String(fileID.prefix(upTo: firstSlash))
+        }
+        return nil
+    }
+}
+
+extension DeclGroupSyntax {
+    fileprivate var typeName: String? {
+        switch self.kind {
+        case .actorDecl, .classDecl, .enumDecl, .structDecl:
+            return self.asProtocol(NamedDeclSyntax.self)?.name.text
+        case .extensionDecl:
+            return self.as(ExtensionDeclSyntax.self)?.extendedType.description
+        default:
+            break
+        }
```

Comment:
Can’t you just do the following here to be compatible with any new types that might be added to `NamedDeclSyntax` in the future?

```suggestion
        if let namedDecl = self.asProtocol(NamedDeclSyntax.self) {
          return namedDecl.name.text
        } else if let extension = self.as(ExtensionDeclSyntax.self) {
          return extension.extendedType.description
        } else {
          return nil
        }
```

Reply:
the macro doesn't support all conformances of `NamedDeclSyntax`, specifically `ProtocolDeclSyntax` is not supported. It's not clear to me that the macro should by default accept future conformances of `NamedDeclSyntax`, since they might be like `protocol`, a type that shouldn't be supported.

Reply:
Oh, I see. I thought you wanted to support all subtypes of `NamedDeclSyntax`. Good with me then. 👍🏽 

Could maybe add a comment though.

Reply:
done